File: backend/app/schemas/system.py
from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.common import PaginationResponseBase, SelectOption
from app.schemas.contact import validate_optional_email, validate_optional_phone_number


class PermissionOption(BaseModel):
    code: str
    name: str
    module_name: str
    description: str


class RoleRecord(BaseModel):
    id: int
    role_code: str
    role_name: str
    scope_name: str
    permissions: list[str]
    user_count: int = 0


class RoleUpsert(BaseModel):
    role_code: str
    role_name: str
    scope_name: str
    permissions: list[str] = Field(default_factory=list)


class RoleListResponse(PaginationResponseBase):
    items: list[RoleRecord]


class SystemUserRecord(BaseModel):
    id: int
    username: str
    full_name: str
    role_code: str
    role_name: str
    department_name: str
    introduction: str | None = None
    email: str | None = None
    phone_number: str | None = None
    account_status: str
    last_login_at: str | None = None


class SystemUserUpsert(BaseModel):
    username: str
    full_name: str
    role_code: str
    department_name: str
    introduction: str | None = None
    email: str | None = None
    phone_number: str | None = None
    account_status: str
    password: str | None = None

    @field_validator("introduction")
    @classmethod
    def normalize_introduction(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = str(value).strip()
        return normalized or None

    @field_validator("email")
    @classmethod
    def validate_email_field(cls, value: str | None) -> str | None:
        return validate_optional_email(value)

    @field_validator("phone_number")
    @classmethod
    def validate_phone_number_field(cls, value: str | None) -> str | None:
        return validate_optional_phone_number(value)

    @model_validator(mode="after")
    def validate_advisor_introduction(self) -> "SystemUserUpsert":
        if self.role_code == "advisor" and not self.introduction:
            raise ValueError("导师角色必须填写介绍")
        return self


class SystemUserListResponse(PaginationResponseBase):
    items: list[SystemUserRecord]


class AuditPolicyRecord(BaseModel):
    id: int
    item: str
    policy: str
    status: str


class AuditPolicyUpsert(BaseModel):
    item: str
    policy: str
    status: str


class AuditPolicyListResponse(PaginationResponseBase):
    items: list[AuditPolicyRecord]


class IntegrationRecord(BaseModel):
    id: int
    name: str
    direction: str
    cadence: str
    status: str
    owner: str


class IntegrationUpsert(BaseModel):
    name: str
    direction: str
    cadence: str
    status: str
    owner: str


class IntegrationListResponse(PaginationResponseBase):
    items: list[IntegrationRecord]


class OperationLogRecord(BaseModel):
    id: int
    operated_at: str
    operator_username: str
    module_name: str
    entity_name: str
    entity_id: str
    action: str
    result: str
    summary: str


class OperationLogListResponse(PaginationResponseBase):
    items: list[OperationLogRecord]


class SyncLogRecord(BaseModel):
    id: int
    source_system: str
    target_system: str
    sync_status: str
    record_count: int
    executed_at: str
    failure_reason: str | None = None


class SyncLogListResponse(PaginationResponseBase):
    items: list[SyncLogRecord]


class NotificationDeliveryLogRecord(BaseModel):
    id: int
    channel: str
    template_code: str | None = None
    recipient: str
    subject: str
    send_status: str
    sent_at: str
    business_key: str | None = None
    triggered_by: str | None = None
    failure_reason: str | None = None


class NotificationDeliveryLogListResponse(PaginationResponseBase):
    items: list[NotificationDeliveryLogRecord]


class PermissionCatalogResponse(BaseModel):
    items: list[PermissionOption]


class DictTypeRecord(BaseModel):
    id: int
    dict_name: str
    dict_type: str
    status: str
    remark: str | None = None
    data_count: int = 0


class DictTypeUpsert(BaseModel):
    dict_name: str
    dict_type: str
    status: str
    remark: str | None = None


class DictTypeListResponse(PaginationResponseBase):
    items: list[DictTypeRecord]


class DictDataRecord(BaseModel):
    id: int
    dict_type: str
    dict_name: str
    label: str
    value: str
    sort_order: int
    status: str
    color_type: str | None = None
    css_class: str | None = None
    remark: str | None = None


class DictDataUpsert(BaseModel):
    dict_type: str
    label: str
    value: str
    sort_order: int = 0
    status: str
    color_type: str | None = None
    css_class: str | None = None
    remark: str | None = None


class DictDataListResponse(PaginationResponseBase):
    items: list[DictDataRecord]


class SystemOptionsResponse(BaseModel):
    account_status_options: list[SelectOption]
    role_scope_options: list[SelectOption]
    integration_direction_options: list[SelectOption]
    integration_cadence_options: list[SelectOption]
    integration_status_options: list[SelectOption]
    audit_status_options: list[SelectOption]
    operation_result_options: list[SelectOption]
    sync_status_options: list[SelectOption]


class BulkDeleteRequest(BaseModel):
    ids: list[int] = Field(default_factory=list)


class BulkActionResponse(BaseModel):
    success_count: int


class SystemArchitecture(BaseModel):
    authentication: str
    database: str
    cache: str
    audit: list[str]
    integrations: list[str]


class SystemStats(BaseModel):
    integration_total: int
    active_integration_total: int
    operation_log_total: int
    sync_failure_total: int
    user_total: int
    role_total: int

File: frontend/src/api/system.ts
import type { BulkActionResponse, PagedResponse, PaginationParams, SelectOption } from './common'
import http from './http'


export type { BulkActionResponse, PagedResponse, PaginationParams, SelectOption } from './common'


export type PermissionOption = {
  code: string
  name: string
  module_name: string
  description: string
}


export type DictTypeRecord = {
  id: number
  dict_name: string
  dict_type: string
  status: string
  remark?: string | null
  data_count: number
}


export type DictTypeUpsert = {
  dict_name: string
  dict_type: string
  status: string
  remark?: string | null
}


export type DictDataRecord = {
  id: number
  dict_type: string
  dict_name: string
  label: string
  value: string
  sort_order: number
  status: string
  color_type?: string | null
  css_class?: string | null
  remark?: string | null
}


export type DictDataUpsert = {
  dict_type: string
  label: string
  value: string
  sort_order: number
  status: string
  color_type?: string | null
  css_class?: string | null
  remark?: string | null
}


export type RoleRecord = {
  id: number
  role_code: string
  role_name: string
  scope_name: string
  permissions: string[]
  user_count: number
}


export type RoleUpsert = {
  role_code: string
  role_name: string
  scope_name: string
  permissions: string[]
}


export type SystemUserRecord = {
  id: number
  username: string
  full_name: string
  role_code: string
  role_name: string
  department_name: string
  phone_number?: string | null
  account_status: string
  last_login_at?: string | null
}


export type SystemUserUpsert = {
  username: string
  full_name: string
  role_code: string
  department_name: string
  phone_number?: string | null
  account_status: string
  password?: string | null
}


export type AuditPolicyRecord = {
  id: number
  item: string
  policy: string
  status: string
}


export type AuditPolicyUpsert = {
  item: string
  policy: string
  status: string
}


export type IntegrationRecord = {
  id: number
  name: string
  direction: string
  cadence: string
  status: string
  owner: string
}


export type IntegrationUpsert = {
  name: string
  direction: string
  cadence: string
  status: string
  owner: string
}


export type OperationLogRecord = {
  id: number
  operated_at: string
  operator_username: string
  module_name: string
  entity_name: string
  entity_id: string
  action: string
  result: string
  summary: string
}


export type SyncLogRecord = {
  id: number
  source_system: string
  target_system: string
  sync_status: string
  record_count: number
  executed_at: string
  failure_reason?: string | null
}


export type SystemStats = {
  integration_total: number
  active_integration_total: number
  operation_log_total: number
  sync_failure_total: number
  user_total: number
  role_total: number
}


export type SystemOptions = {
  account_status_options: SelectOption[]
  role_scope_options: SelectOption[]
  integration_direction_options: SelectOption[]
  integration_cadence_options: SelectOption[]
  integration_status_options: SelectOption[]
  audit_status_options: SelectOption[]
  operation_result_options: SelectOption[]
  sync_status_options: SelectOption[]
}


export function getSystemStats() {
  return http.get<SystemStats>('/system/stats')
}


export function getSystemOptions() {
  return http.get<SystemOptions>('/system/options')
}


export function getPermissionCatalog() {
  return http.get<{ items: PermissionOption[] }>('/system/permissions')
}


export function listDictTypes(params?: PaginationParams & {
  keyword?: string
  status?: string
}) {
  return http.get<PagedResponse<DictTypeRecord>>('/system/dict-types', { params })
}


export function createDictType(payload: DictTypeUpsert) {
  return http.post<DictTypeRecord>('/system/dict-types', payload)
}


export function updateDictType(id: number, payload: DictTypeUpsert) {
  return http.put<DictTypeRecord>(`/system/dict-types/${id}`, payload)
}


export function deleteDictType(id: number) {
  return http.delete(`/system/dict-types/${id}`)
}


export function listDictData(params?: PaginationParams & {
  keyword?: string
  dict_type?: string
  status?: string
}) {
  return http.get<PagedResponse<DictDataRecord>>('/system/dict-data', { params })
}


export function createDictData(payload: DictDataUpsert) {
  return http.post<DictDataRecord>('/system/dict-data', payload)
}


export function updateDictData(id: number, payload: DictDataUpsert) {
  return http.put<DictDataRecord>(`/system/dict-data/${id}`, payload)
}


export function deleteDictData(id: number) {
  return http.delete(`/system/dict-data/${id}`)
}


export function listSystemUsers(params?: PaginationParams & {
  keyword?: string
  role_code?: string
  account_status?: string
  department_name?: string
}) {
  return http.get<PagedResponse<SystemUserRecord>>('/system/users', { params })
}


export function createSystemUser(payload: SystemUserUpsert) {
  return http.post<SystemUserRecord>('/system/users', payload)
}


export function updateSystemUser(id: number, payload: SystemUserUpsert) {
  return http.put<SystemUserRecord>(`/system/users/${id}`, payload)
}


export function deleteSystemUser(id: number) {
  return http.delete(`/system/users/${id}`)
}


export function batchDeleteSystemUsers(ids: number[]) {
  return http.post<BulkActionResponse>('/system/users/batch-delete', { ids })
}


export function listRoles(params?: PaginationParams & {
  keyword?: string
  scope_name?: string
  permission?: string
}) {
  return http.get<PagedResponse<RoleRecord>>('/system/roles', { params })
}


export function createRole(payload: RoleUpsert) {
  return http.post<RoleRecord>('/system/roles', payload)
}


export function updateRole(id: number, payload: RoleUpsert) {
  return http.put<RoleRecord>(`/system/roles/${id}`, payload)
}


export function deleteRole(id: number) {
  return http.delete(`/system/roles/${id}`)
}


export function batchDeleteRoles(ids: number[]) {
  return http.post<BulkActionResponse>('/system/roles/batch-delete', { ids })
}


export function listAuditPolicies(params?: PaginationParams & {
  keyword?: string
  status?: string
}) {
  return http.get<PagedResponse<AuditPolicyRecord>>('/system/audit-policies', { params })
}


export function createAuditPolicy(payload: AuditPolicyUpsert) {
  return http.post<AuditPolicyRecord>('/system/audit-policies', payload)
}


export function updateAuditPolicy(id: number, payload: AuditPolicyUpsert) {
  return http.put<AuditPolicyRecord>(`/system/audit-policies/${id}`, payload)
}


export function deleteAuditPolicy(id: number) {
  return http.delete(`/system/audit-policies/${id}`)
}


export function batchDeleteAuditPolicies(ids: number[]) {
  return http.post<BulkActionResponse>('/system/audit-policies/batch-delete', { ids })
}


export function listIntegrations(params?: PaginationParams & {
  keyword?: string
  status?: string
  direction?: string
}) {
  return http.get<PagedResponse<IntegrationRecord>>('/system/integrations', { params })
}


export function createIntegration(payload: IntegrationUpsert) {
  return http.post<IntegrationRecord>('/system/integrations', payload)
}


export function updateIntegration(id: number, payload: IntegrationUpsert) {
  return http.put<IntegrationRecord>(`/system/integrations/${id}`, payload)
}


export function deleteIntegration(id: number) {
  return http.delete(`/system/integrations/${id}`)
}


export function batchDeleteIntegrations(ids: number[]) {
  return http.post<BulkActionResponse>('/system/integrations/batch-delete', { ids })
}


export function listOperationLogs(params?: PaginationParams & {
  keyword?: string
  module_name?: string
  result?: string
}) {
  return http.get<PagedResponse<OperationLogRecord>>('/system/operation-logs', { params })
}


export function listSyncLogs(params?: PaginationParams & {
  keyword?: string
  sync_status?: string
  source_system?: string
}) {
  return http.get<PagedResponse<SyncLogRecord>>('/system/sync-logs', { params })
}
